### PR TITLE
FIX: UnsavedRelationList first/last to return null if list is empty (fixes #11083)

### LINKED ISSUE #11083: UnsavedRelationList::first() / last() return false instead of null when list is empty
Pull request: https://github.com/silverstripe/silverstripe-framework/pull/11085

## Affected Version

5.x

## Description

`UnsavedRelationList::first()` and `UnsavedRelationList::last()` return `false` instead of `null` when the list is empty. This is inconsistent with `DataList` and co, which return `null`. Cause is the use of `reset()` and `end()`:

https://github.com/silverstripe/silverstripe-framework/blob/809f9e7ae06e4e3999ef2670aa322aacdddb7def/src/ORM/UnsavedRelationList.php#L238-L248

https://github.com/silverstripe/silverstripe-framework/blob/809f9e7ae06e4e3999ef2670aa322aacdddb7def/src/ORM/UnsavedRelationList.php#L255-L262

I suppose there’s a minor BC risk to changing this, open to feedback on whether this is appropriate for a patch release.

### PR DESCRIPTION
**Pull request by lozcalver**
Issue #11083

### CODE REVIEW

File: src/ORM/UnsavedRelationList.php [hunk 1]
```diff
@@ -239,11 +239,11 @@ public function getIDList()
      */
     public function first()
     {
-        $item = reset($this->items);
+        $item = reset($this->items) ?: null;
         if (is_numeric($item)) {
             $item = DataObject::get_by_id($this->dataClass, $item);
         }
-        if (!empty($this->extraFields[key($this->items)])) {
+        if ($item && !empty($this->extraFields[key($this->items)])) {
```
Comment (lozcalver) [inline src/ORM/UnsavedRelationList.php:246]: Not actually part of this fix but seemed like sensible defensive coding. Happy to remove or break out into a separate commit if preferred

File: src/ORM/UnsavedRelationList.php [hunk 2]
```diff
@@ -256,8 +256,11 @@ public function first()
      */
     public function last()
     {
-        $item = end($this->items);
-        if (!empty($this->extraFields[key($this->items)])) {
+        $item = end($this->items) ?: null;
+        if (is_numeric($item)) {
+            $item = DataObject::get_by_id($this->dataClass, $item);
```
Comment (lozcalver) [inline src/ORM/UnsavedRelationList.php:261]: Pushed a second commit for this bugfix. tldr; `UnsavedRelationList` items array can contain both full objects (for unsaved records) or just IDs (for already existing records). This code already exists in `first()`, I think `last()` was just overlooked.

Covered by the test already added for the main fix (that’s how I noticed this!): https://github.com/silverstripe/silverstripe-framework/actions/runs/7085966036/job/19283271493?pr=11085

File: src/ORM/UnsavedRelationList.php (hunk 1 above)
Comment (GuySartorelli) [reply src/ORM/UnsavedRelationList.php:246]: Seems sensible.

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**